src/scripts/http/cachedHttp.ts

```diff
@@ -0,0 +1,121 @@
+import {ResponsePackage} from "../responsePackage";
+
+import {Storage} from "../storage/storage";
+
+export interface TimeStampedData {
+	data: any;
+	lastUpdated: number;
+}
+
+export type GetResponseAsync = () => Promise<ResponsePackage<string>>;
+
+/**
+ * Allows the creation of HTTP GET requests to retrieve data, as well as caching of the result.
+ */
+export class CachedHttp {
+	private cache: Storage;
+
+	constructor(cache: Storage) {
+		this.cache = cache;
+	}
+
+	/**
+	 * Given a key, checks the cache for a fresh copy of that value. If the cached value does
+	 * not exist, or is not fresh, fetches a fresh copy of the data from the specified endpoint.
+	 */
+	public getAndCacheFreshValue(key: string, getRemoteValue: GetResponseAsync, updateInterval: number): Promise<TimeStampedData> {
+		if (!key) {
+			throw new Error("key must be a non-empty string, but was: " + key);
+		} else if (!getRemoteValue) {
+			throw new Error("getRemoteValue must be non-undefined");
+		} else if (updateInterval < 0) {
+			updateInterval = 0;
+		}
+
+		return new Promise<TimeStampedData>((resolve, reject) => {
+			let value = this.cache.getValue(key);
+			let keyIsPresent = !!value;
+			if (keyIsPresent) {
+				let valueAsJson: TimeStampedData;
+				try {
+					valueAsJson = JSON.parse(value);
+				} catch (e) {
+					reject({ error: e });
+				}
+
+				let valueHasExpired = CachedHttp.valueHasExpired(valueAsJson, updateInterval);
+
+				if (!valueHasExpired) {
+					resolve(valueAsJson);
+					return;
+				}
+			}
+
+			this.getAndCacheRemoteValue(key, getRemoteValue).then((timeStampedData) => {
+				resolve(timeStampedData);
+			}, (error) => {
+				reject(error);
+			});
+		});
+	}
+
+	public getAndCacheRemoteValue(key: string, getRemoteValue: GetResponseAsync): Promise<TimeStampedData> {
```

Comment:
I wish there was a middle ground between `public` and `private` in TS... because this implementation (and the one in `ClipperCachedHttp`) aren't implementations we actually want to be using in our codebase, I imagine. So having them `public`... feels too tempting/easy for someone to lose track of some of the abstraction going on here and just start calling it (in all of its log-less, throwing glory ;P)

I get the pattern you're going for, but could we perhaps just make some of these low-level implementations (that are currently only being directly used in tests) throw an Not Implemented error and then condense all desired implementation behavior into `ClipperData`?